include/SampleThumbnail.h

```diff
@@ -0,0 +1,114 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_SAMPLE_THUMBNAIL_H
+#define LMMS_SAMPLE_THUMBNAIL_H
+
+#include <QPainter>
+#include <QRect>
+#include <memory>
+#include "Sample.h"
+#include "lmms_export.h"
+
+namespace lmms {
+
+/**
+   Insert this into your class when you want to implement
+   thumbnails. This is optional, but without this class, there will be no
+   indication that this thumbnail list is being used and it will be destroyed
+   when you generate thumbnails somewhere else, and have to be regenerated
+   before being rendered, which is slower than rendering the sample directly.
+*/
+class LMMS_EXPORT SampleThumbnail
+{
+public:
+	struct Bit
+	{
+		Bit() = default;
+
+		Bit(float min, float max): min(min), max(max) {}
+
+		Bit(const SampleFrame& frame)
+			: min(std::min(frame.left(), frame.right()))
+			, max(std::max(frame.left(), frame.right()))
+		{}
+
+		Bit operator+(const Bit& other) const
+		{
+			return Bit(std::min(min, other.min), std::max(max, other.max));
+		}
+
+		Bit operator+(const SampleFrame& frame) const
+		{
+			return *this + Bit{frame};
+		}
+
+		float min =  std::numeric_limits<float>::max();
+		float max = -std::numeric_limits<float>::max();
+	};
+
+	struct VisualizeParameters
+	{
+		float amplification = 1.0f; //!< The amount of amplification to apply to the waveform.
+		bool reversed = false;		//!< Determines if the waveform is drawn in reverse or not.
+
+		/*
+			You can set these members when there's no easy way to calculate sampRect.
+			These will stretch the viewed region to the dimensions of the clipRect.
+		*/
+		float sampleStart = 0.0f; //!< Where the sample begins for drawing.
+		float sampleEnd = 1.0f;	  //!< Where the sample ends for drawing.
+
+		QRect sampRect = QRect(); 	//!< Dimensions of the fully rendered sample on the Song editor; Can move around. = clipRect when null.
+		QRect clipRect;				//!< Region that the sample will be rendered into; Fixed in place.
+		QRect drawRect = QRect(); 	//!< Region of clipRect that will be drawn into (the update region). = clipRect when null.
+	};
+
+	using Thumbnail = std::vector<Bit>;
+	using ThumbnailCache = std::vector<Thumbnail>;
+
+	SampleThumbnail() = default;
+	SampleThumbnail(const Sample& sample);
```

Comment:
Sorry, but I was hoping that we would stop passing in a `Sample` object altogether. `Sample` is really a class that wraps a buffer around other stuff necessary for playback. The only thing you really need to pass here is `SampleFrame*` and `size_t` for the buffer itself and the size respectively.

Reply:
We're only passing references around no? It shouldn't copy stuff into the function. If it actually improves performance then I guess I'll change it to something like `(const SampleFrame* buffer, size_t sampleSize)`

Reply:
The performance impact involves cache misses, which is probably negligible here. I just don't think it's necessary since `Sample` is really a class that handles playback: you just need the audio buffer here, nothing else.

Reply:
I should have really named `Sample` something like `SamplePlayer` to avoid confusion.

Reply:
Wait, the `selectFromGlobalThumbnailMap` currently depends on the name of the sample file, so I don't think I can replace that for now.